File: L1/cpp/test/testPrimesGenerator.cpp
#include <primesGenerator.hpp>

#include <chrono>
#include <iostream>
#include <cassert>

void testPrimesGenerator() noexcept(true);

static void testPrimesGeneratorException(const int n, const std::vector<size_t>& primeIndexes) noexcept(true)
{
    try
    {
        const PrimesGenerator primesGenerator(n);
        for (const size_t& primeIndex : primeIndexes)
        {
            const int prime = primesGenerator[primeIndex];
            (void)prime;
        }

        assert(false && "Exception should be raised");
    }
    catch (const std::invalid_argument& e)
    {
        (void)e;
    }
}

static void testPrimesGeneratorNormalCase(const int n, const std::vector<size_t>& primeIndexes, const std::vector<int> expectedPrimes) noexcept(true)
{
    const PrimesGenerator primesGenerator(n);
    for (size_t i = 0; i < primeIndexes.size(); ++i)
        assert(primesGenerator[primeIndexes[i]] == expectedPrimes[i]);
}

static void testPrimesGeneratorPerformance(const int n, const std::vector<size_t>& primeIndexes) noexcept(true)
{
    const auto startTime = std::chrono::high_resolution_clock::now();

    const PrimesGenerator primesGenerator(n);
    for (const size_t& primeIndex : primeIndexes)
    {
        const int prime = primesGenerator[primeIndex];
        (void)prime;
    }

    const auto finishTime = std::chrono::high_resolution_clock::now();

    std::cout << "PrimesGenerator(" << n << ") took: " << std::chrono::duration_cast<std::chrono::milliseconds>(finishTime - startTime).count() << "ms" << std::endl;
}

void testPrimesGenerator() noexcept(true)
{
    testPrimesGeneratorException(-100, std::vector<size_t>{});
    testPrimesGeneratorException(0, std::vector<size_t>{});
    testPrimesGeneratorException(1, std::vector<size_t>{});
    testPrimesGeneratorException(2, std::vector<size_t>{0, static_cast<size_t>(-1), 1, 2, 100, static_cast<size_t>(-100)});
    testPrimesGeneratorException(100, std::vector<size_t>{0, static_cast<size_t>(-1), 1, 2, 100, static_cast<size_t>(-100)});
    testPrimesGeneratorException(1000, std::vector<size_t>{0, static_cast<size_t>(-1), 1, 2, 100, static_cast<size_t>(-100)});

    testPrimesGeneratorNormalCase(2, std::vector<size_t>{0}, std::vector<int>{2});
    testPrimesGeneratorNormalCase(100, std::vector<size_t>{0, 3, 7, 10}, std::vector<int>{2, 7, 19, 31});

    testPrimesGeneratorPerformance(10 * 1000, std::vector<size_t>{0, 100, 1000});
    testPrimesGeneratorPerformance(1 * 1000 * 1000, std::vector<size_t>{0, 100, 1000});
    testPrimesGeneratorPerformance(1 * 1000 * 1000, std::vector<size_t>{0, 100, 1000, 78497});
    testPrimesGeneratorPerformance(1 * 1000 * 1000 * 1000, std::vector<size_t>{0, 100, 1000});
    testPrimesGeneratorPerformance(1 * 1000 * 1000 * 1000, std::vector<size_t>{0, 100, 1000, 100000});
}

File: L1/java/TestPrimesGenerator.java
public class TestPrimesGenerator
{
    private TestPrimesGenerator() throws InstantiationError
    {
        throw new InstantiationError("This is a static class!");
    }

    private static void testPrimesGeneratorException(final int n, final int [] primeIndexes)
    {
        try
        {
            final PrimesGenerator primesGenerator = new PrimesGenerator(n);
            for (final int index : primeIndexes)
            {
                final int prime = primesGenerator.getPrime(index);
            }

            assert false : "Exception should be raised";

        }
        catch (final IllegalArgumentException e)
        {

        }
    }

    private static void testPrimesGeneratorNormalCase(final int n, final int[] primeIndexes, final int[] expectedPrimes)
    {
        final PrimesGenerator primesGenerator = new PrimesGenerator(n);
        for (int i = 0; i < primeIndexes.length; ++i)
        {
            final int prime = primesGenerator.getPrime(primeIndexes[i]);
            assert prime == expectedPrimes[i] : "prime should be equals with expected prime";
        }
    }

    private static void testPrimesGeneratorPerformance(final int n, final int[] primeIndexes)
    {
        final long startTime = System.currentTimeMillis();

        final PrimesGenerator primesGenerator = new PrimesGenerator(n);
        for (int i = 0; i < primeIndexes.length; ++i)
        {
            final int prime = primesGenerator.getPrime(primeIndexes[i]);
        }

        final long stopTime = System.currentTimeMillis();
        System.out.println("primesGenerator(" + n + ") took: " + (stopTime - startTime) + "ms");
    }

    public static void testPrimesGenerator()
    {
        testPrimesGeneratorException(-100, new int[] {});
        testPrimesGeneratorException(-1, new int[] {});
        testPrimesGeneratorException(0, new int[] {});
        testPrimesGeneratorException(1, new int[] {});
        testPrimesGeneratorException(2, new int[] {0, -1, 1, 2, 100, -100});
        testPrimesGeneratorException(100, new int[] {0, -1, 1, 2, 100, -100});
        testPrimesGeneratorException(1000, new int[] {0, -1, 1, 2, 100, -100});

        testPrimesGeneratorNormalCase(2, new int[] {0}, new int[] {2});
        testPrimesGeneratorNormalCase(100, new int[] {0, 3, 7, 10}, new int[] {2, 7, 19, 31});

        testPrimesGeneratorPerformance(10 * 1000, new int[] {0, 100, 1000});
        testPrimesGeneratorPerformance(1 * 1000 * 1000, new int[] {0, 100, 1000});
        testPrimesGeneratorPerformance(1 * 1000 * 1000, new int[] {0, 100, 1000, 78497});
        testPrimesGeneratorPerformance(1 * 1000 * 1000 * 1000, new int[] {0, 100, 1000}); // java -Xmx16G
        testPrimesGeneratorPerformance(1 * 1000 * 1000 * 1000, new int[] {0, 100, 1000, 100000});
    }
}
